[오목 1, 2단계] 조이 미션 제출합니다.

gahyunkim opened the pull request:
안녕하세요,코니!
이번 레벨1 마지막 단계 오목미션을 리뷰 받게 된 조이입니다.
리뷰해주시는 내용을 바탕으로 더 열심히 1,2단계 리팩토링 해보도록 하겠습니다.
감사합니다!

# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

<!--
학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
-->

- 오목판의 상태가 계속해서 변화하기 때문에 불변성을 지키도록 집중해서 구현해보았습니다. 불변성이 잘 지켜지고 있는지, 불변성과 관련해서 리뷰해주시면 좋을 것 같습니다!
- 객체의 역할과 책임과 관련해서 더 많이 공부해보고 싶습니다. 현재 코드를 바탕으로 객체가 어떤지 생각해볼 수 있는 질문을 많이 남겨주시면 좋을 것 같습니다.

gahyunkim committed:
docs: Readme 기능 요구 사항 추가

gahyunkim committed:
test: BoardTest 테스트 코드 구현

gahyunkim committed:
feat: 오목판 생성 및 돌 착수 코드 구현

gahyunkim committed:
docs: Readme 요구사항 완료 체크

gahyunkim committed:
feat: Direction enum class 생성

gahyunkim committed:
test: 오목 확인을 위한 테스트 코드 구현

gahyunkim committed:
feat: 오목 검증 로직 구현

gahyunkim committed:
docs: Readme 요구사항 완료 체크

gahyunkim committed:
refactor: 오목판 자료구조 Map으로 개선

gahyunkim committed:
test: BoardTest 테스트 코드 수정

gahyunkim committed:
refactor: nextStoneState 코드 수정

gahyunkim committed:
feat: view 입출력 코드 구현

gahyunkim committed:
feat: Controller, Application 코드 구현

gahyunkim committed:
test: BoardTest 테스트 코드 수정

gahyunkim committed:
refactor: Board 인스턴스 변수 변경

gahyunkim committed:
refactor: 가시성 변경자 수정

gahyunkim committed:
feat: 마지막 돌 출력 코드 구현

gahyunkim committed:
docs: Readme 기능 요구 사항 추가

gahyunkim committed:
refactor: isOmok을 isLastStoneOmok으로 변경

gahyunkim committed:
test: isOmok을 isLastStoneOmok으로 변경 후 테스트 코드 구현

gahyunkim committed:
feat: 오목 우승자 판정 로직 적용

gahyunkim committed:
feat: 오목 우승자 출력 코드 구현

gahyunkim committed:
feat: 렌주룰 라이브러리 의존성 추가

gahyunkim committed:
feat: Position을 렌주룰 라이브러리 Point로 변경하는 코드 구현

gahyunkim committed:
feat: 흑돌 렌주룰 적용 코드 구현

gahyunkim committed:
feat: 흑돌 금수 좌표 입력 시 재시도 관련 로직 구현

gahyunkim committed:
docs: Readme 요구사항 완료 체크

gahyunkim committed:
feat: OmokRule 클래스 구현

gahyunkim committed:
test: OmokRuleTest로 테스트 코드 이동

gahyunkim committed:
test: BoardTest 테스트 코드 일부 삭제

gahyunkim committed:
feat: Direction을 라인별로 묶는 코드 구현

gahyunkim committed:
feat: Stone에서 Pair의 값을 받아오는 부생성자 구현

gahyunkim committed:
refactor: omok룰 관련 로직을 OmokRule로 이동

gahyunkim committed:
refactor: 변수 및 함수명 변경, parsing 코드 이동

gahyunkim committed:
test: Parsing 테스트 코드 구현

gahyunkim committed:
feat: Parsing 코드 구현

gahyunkim committed:
feat: 에러 처리문 출력 코드 구현

gahyunkim committed:
feat: 에러 출력을 outputView로 위임

gahyunkim committed:
feat: 텍스트 및 숫자 상수화

gahyunkim committed:
feat: 패키지 역할별로 구분

gahyunkim committed:
chore: gitkeep 삭제

gahyunkim committed:
refactor: 최댓값 반환 로직 수정

gahyunkim committed:
refactor: 오목 여부 반환값 수정

gahyunkim committed:
refactor: 코틀린 코드 컨벤션 반영

gahyunkim committed:
fix: 백돌이 흑돌의 금수 위치에 착수할 수 없는 오류 수정

lee-ji-hoon commented on src/main/kotlin/omok/model/stone/position/Position.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    constructor(coordinateText: String) : this(
+        row = Row((coordinateText.substring(1).toIntOrNull() ?: throw IllegalArgumentException(ERROR_ROW_NUM)) - 1),
```
부 생성자를 활용해 보려는 시도는 좋았습니다.

단 지금 이 로직이 Model이 해야 할 상황인지 생각해 보면 좋을 것 같습니다.

전반적으로 무조건 View가 `A1` 이런 형태로 입력이 될 것이라고 판단을 하는데 Model입장에서는 내가 어디에 놓여 있는지 `좌표`가 중요하지 않을까요?

gahyunkim replied:
[refactor: parsing 위치 이동 및 boarSize 확장성 적용](https://github.com/woowacourse/kotlin-omok/pull/115/commits/e6e8b538474369789e94fa92a96807858bdf0b79)

Model 입장에서는 좌표만 알면되고 사용자가 입력한 'H8'과 같은 값은 알 필요가 없다고 생각합니다. 그래서 사용자가 입력한 값에 대해서 inputView내에서 `fun parse()`를 만들어주어 Parsing을 진행하고, Parsing된 값을 반환하도록 해서 컨트롤러 내에서 사용할 수 있도록 구현해주었습니다!

다만, 그렇게 되면 뷰에서 파싱과 동시에 값에 대한 검증을 진행하게 되기 때문에 Position이나 Row, Col에서 별도의 검증이 필요하지 않아지는 것인지에 대해서 의문이 생깁니다. 

제 생각엔 뷰에서 들어오는 값에 대해서 도메인은 뷰를 무조건적으로 믿어서는 안된다고 생각해서 도메인 내에서도 검증을 해줘야 한다고 생각합니다. 또한, 테스트 코드를 만들때도 테스트 코드 내에서 제대로 돌아가는 지 확인하기 위해서 도메인에서도 검증을 해야한다고 생각합니다.

이에 대해서 코니는 어떻게 생각하시는지 궁금합니다!

lee-ji-hoon replied:
네 좋습니다. 어떤 의미로 말씀하신 것인지는 이해가 되네요.

```
도메인은 뷰를 무조건적으로 믿으면 안 된다.
```

라는 내용이 핵심 같은데요. 그러면 어디까지 믿어야 하는 걸까요? 안 믿을 거면 아예 `-` 로 들어오지는 않았는지도 확인을 했어야 할거 같은데 그렇다면 `Row` / `Col` 에서도 동일하게 검증을 했어야겠네요.

`Model` 이 만들어질 때는 어떠한 역할을 해야 하고, 무슨 비즈니스 로직을 담아 둘 것인가를 메인으로 생각해 보면 `Row` / `Col` 이 `Board Size` 에 굳이 의존을 해야 할까? 싶네요. 그리고 실제로 지금도 View에 `BoardSize` 를 갖고 와서 쓰고 있기도 하고요.

그런 관점에서 봤을 때 이미 parse를 하고 있는데 board의 size를 알아서 view가 처리하는 게 그리 이상한가? 를 고민해 보면 저는 아니라고 생각이 드는데 조이의 추가적인 의견이 궁금하네요.

gahyunkim replied:
[refactor: Position에서의 검증 삭제](https://github.com/woowacourse/kotlin-omok/pull/115/commits/03a8c069636bde9397eb405db7093fce6b5dc9c2)
[refactor: boardSize 변경으로 인한 코드 수정 및 검증 추가](https://github.com/woowacourse/kotlin-omok/pull/115/commits/c8a998fafec80b72b4a0860495c525c7c09b261d)

parsing하는 과정에서 어차피 보드의 크기를 알고 이를 바탕으로 검증을 해주고 있기 때문에 코니의 의견처럼 view에서 검증을 해주는 것도 크게 이상하지 않다고 생각합니다!

>Model 이 만들어질 때는 어떠한 역할을 해야 하고, 무슨 비즈니스 로직을 담아 둘 것인가를 메인으로 생각해 보면 Row / Col 이 Board Size 에 굳이 의존을 해야 할까?

라는 코니의 말을 듣고 나니, 굳이 Row와 Col이 BoardSize를 알고 있어야 할 필요가 없다는 생각이 드네요. 결국 안에서 가져야 하는 값은 Int값이고 이들은 굳이 정해진 범위 내의 값을 가져야한다는 제약을 두지 않아도 될 것 같습니다!

lee-ji-hoon commented on src/main/kotlin/omok/model/stone/position/Position.kt:
```diff
@@ -0,0 +1,38 @@
+        private const val MIN_COL_CHAR = 'A'
+        private const val MAX_COL_CHAR = 'O'
+        private const val MIN_RANGE = 0
+        private const val MAX_RANGE = 14
```
`BOARD_SIZE`는 Border에 있고, `MAX_RANGE`는 Position에 있네요. 

오목의 경우 무조건 15x15만 있지는 않을 거 같은데 만약 `100x100` 으로 변경이 된다면 수정 사항이 한 곳이 아니라 여러 곳이 되게 되는데 이러면 실수하기 쉽지 않을까요?

gahyunkim replied:
[refactor: parsing 위치 이동 및 boarSize 확장성 적용](https://github.com/woowacourse/kotlin-omok/pull/115/commits/e6e8b538474369789e94fa92a96807858bdf0b79)

BOARD_SIZE를 15로 고정된 상수로 만들어 사용하고 있는데, BOARD_SIZE를 outputView, InputView, Position에서 사용하고 있습니다.

말씀해주신것처럼 15*15가 아닌 경우가 있다고 생각할 때, 아래의 코드가 0과 14로 고정되어있기 때문에 확장성을 고려한다면 문제가 생길 수 있을 것 같습니다. 
```kotlin
private const val MIN_RANGE = 0
private const val MAX_RANGE = 14
```

그래서 해당하는 코드들을 BOARD_SIZE가 변경되면 자연스럽게 적용될 수 있도록 구현해주었습니다.

다만, 여기서 고민인 부분은 BOARD_SIZE 상수가 변경될 때 여러 곳을 변경하지 않기 위해서 private이 아닌 public을 사용하게 되었는데 과연 BOARD_SIZE 상수를 public으로 사용하는 것이 맞나? 라는 점입니다.

엘레강트 오브젝트가 저자의 개인적인 소신이 담겨있는 책이기 때문에 무조건적으로 인정하는 것은 아니지만, 아직 소신이 없는 저로서는 엘강오의 이야기가 옳은 것인지 혹은 좋은 방향인지 고민하게 됩니다...

제가 고민하게 된 이유는 엘강오에서 아래와 같이 '퍼블릭 상수를 사용하지마세요' 라는 부분이 있기 때문입니다. 
> 객체들은 어떤 것도 공유해서는 안됩니다. 대신 독립적이어야 하고 '닫혀 있어야' 합니다. 간단하게 말해서 상수를 이용한 공유 매커니즘은 캡슐화와 객체지향적인 사고 전체를 부정하는 일입니다.

BOARD_SIZE라는 여기저기서 사용되는 값에 대해서 public으로 구현해주는 것에 대해서 코니는 어떻게 생각하시나요??

gahyunkim replied:
BOARD_SIZE를 public이 아니라 Board 내에서 size를 가지고 있다고 생각하고, boardDimensions라는 data class를 새로 만들어서 사용해보는 것도 시도해보았지만, 계속해서 InputView, OutputView에서 의존성이 발생하는 것을 확인할 수 있었습니다..

```kotlin
data class BoardDimensions(
    val width: Int,
    val height: Int
)
```

아니면, get()을 이용해서 board내에서 width와 height을 받아오려 했지만, List<List<StoneState>>를 이용해서 해당하는 값을 먼저 크기에 맞게 초기화 한다면 가능하겠지만 현재 Map을 이용하고 있기 때문에 해당 부분은 구현이 어려울 것 같았습니다...

제대로 고민하면서 구현하고 있는지는 모르겠지만 일단, 고민한 부분에 대해서 첨부해둡니다!

lee-ji-hoon replied:
> 객체들은 어떤 것도 공유해서는 안됩니다. 대신 독립적이어야 하고 '닫혀 있어야' 합니다. 간단하게 말해서 상수를 이용한 공유 메커니즘은 캡슐화와 객체지향적인 사고 전체를 부정하는 일입니다.

아직 안 읽어 본 책이긴 하지만 재미있는 내용이네요.

말씀하신 대로 public으로 구현을 하는 것은 지양하는 편이긴 합니다. 그런데 지금 구현하신 부분에서 `15`라는 `BOARD_SIZE` 값이 outputView에 존재해야 했던 이유는 살짝 이해가 안 되긴 하네요. 만약 View가 바뀌었을 때 Controller랑 Model에 무슨 일이 일어날지를 고민해보면 좋겠네요.

lee-ji-hoon replied:
> BOARD_SIZE를 public이 아니라 Board 내에서 size를 가지고 있다고 생각하고, boardDimensions라는 data class를 새로 만들어서 사용해보는 것도 시도해보았지만, 계속해서 InputView, OutputView에서 의존성이 발생하는 것을 확인할 수 있었습니다..

네 좋습니다 추가적인 고민을 잘해주셨는데요.

`Board` 라는 Model을 봤을 때 오히려 `row` `col` 즉 Size를 init으로 갖고 갔을 때 어떤 부분 때문에 Map을 이용하기 때문에 불가능하다고 생각했는지가 궁금하네요.

> 만약 Board에 getWidth / getHeight을 구현하는 형태였다면, 말씀하신대로 public한 상수도 없었을 것이고, View에 출력할 때도 해당 width hieght만 넘기면 출력도 무난했을 것 같네요.

gahyunkim replied:
[feat: BoardDimensions data class 생성](https://github.com/woowacourse/kotlin-omok/pull/115/commits/13efb85148c1884e985b44adac32303a67bef916)
[refactor: boardSize 변경 및 black,whiteStones 필터링 추가로 인한 model 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/94b6281ae84796d8fa811a054d2a8f26924d5197)

View가 BOARD_SIZE라는 상수를 가지고 있었고 public상수로 만들어줬기 때문에 모델 내에서도 해당하는 크기를 사용하기 위해서 모델이 뷰를 알게되는 상황이 이루어졌습니다. 
=> 결론적으로는 public 상수를 사용하지 않는 것이 좋겠다는 생각이 들었습니다...

이전에 `List<List<StoneState>>`를 사용했을 때는 결국 초기화를 통해서 가로 세로의 길이를 알고있어야 했기 때문에 Map을 사용할 때와는 달리 쉽게 가로,세로의 길이를 얻을 수 있다고 생각했습니다. 

하지만, Map을 사용하게 되면 들어오는 stones들만 알고 있고 size와는 무관하기 때문에 board의 size를 쉽게 얻어오기 어려울 것이라는 생각을 했습니다. 
=> 사실 너무 단순하게 size를 얻으려고 생각했던 것 같습니다..

리팩토링을 하면서 BoardDimensions를 다시 시도해봤습니다. 
```kotlin
class Board private constructor(
    val stonesMap: Map<Position, StoneColor> = emptyMap(),
    private val dimensions: BoardDimensions,
) {
    fun getWidth() = dimensions.width

    fun getHeight() = dimensions.height
    ...
}
```
위와 같이 다시 BoardDimensions를 이용해서 구현해주면서 Board만 알면 결국 내부에서 get을 이용해 가로,세로의 길이를 얻을 수 있도록 구현해주었습니다!

lee-ji-hoon commented on src/main/kotlin/omok/model/stone/StoneState.kt:
```diff
@@ -0,0 +1,7 @@
+    BLACK,
+    WHITE,
+    NONE,
+}
```
State라는 것을 배웠기에 시도해 본 것은 좋았습니다. 단 `StoneState`라는 이름을 봤을 때 `돌의 상태`라고 처음에는 느껴져 있는데 막상 보니 색상과 `NONE`이 존재하네요.

`돌의 상태`라는 것에 집중을 해봤을 때 이것이 진짜 상태인지 고민을 해보면 좋을 것 같아요. 단순 `돌` 을 구분하기 위해서 사용된 것으로 밖에 느껴지지 않습니다.

> 그리고 상태를 기반으로 했을 때 NONE은 무슨 상태를 의미한다고 생각하시나요?

gahyunkim replied:
[refactor: StoneState를 StoneColor로 변경](https://github.com/woowacourse/kotlin-omok/pull/115/commits/029c66183af6708208d478fe866416fa86647fa0)
[refactor: StoneColor로 변경으로 인한 model 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/3195e41d868d0b1738fc40e887cc9ad022067c45)
[test: StoneColor로 변경으로 인한 test 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/40cad019b4bddaa8d8329326fb2ecfd801f9aac0)
[refactor: StoneColor로 변경으로 인한 view 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/e28c440d5aa6e01e640b090b22144c4141c4c4b0)

처음에는 `List<List<StoneState>>`를 stonesMap으로 만들어서 오목판 내의 모든 포인트들의 상태가 NONE이 되도록 초기화해서 만들어주다보니 위와 같은 enum class가 등장하게 되었던 것 같습니다.

하지만, 이렇게 되면 모든 좌표를 StoneState로 초기화해줘야 한다는 생각이 들어서 Map을 이용해서 Map<Position, StoneState>와 같은 식으로 변경해주다보니 상태에 관한 목적이 약간 변경되어서 이런 코드가 등장하게 된 것 같습니다!

그래서 StoneState enum클래스 명을 StoneColor로 변경해주었습니다. 
`돌의 상태`라기 보다는 돌의 색을 통해서 해당하는 위치에 돌의 색이 있는지 없는지 혹은 Black인지 White인지 확인할 수 있도록 구현하는 것이 더 옳은 방향이라는 생각이 들었습니다!

lee-ji-hoon commented on src/main/kotlin/omok/model/stone/position/Position.kt:
```diff
@@ -0,0 +1,38 @@
+
+import rule.wrapper.point.Point
+
+data class Position(
```
전반적으로 `Model` 들이 외부 라이브러리를 많이 의존하고 있습니다.

만약 라이브러리가 변경이 되면 `Position`이라는 `Model` 자체에 영향을 끼치게 되는데요. 

순수한 `Model` 들을 만들어보는 연습을 해보면 어떨까요?

gahyunkim replied:
[feat: Point 라이브러리 의존을 줄이기 위한 확장함수 구현](https://github.com/woowacourse/kotlin-omok/pull/115/commits/c549cc683775226d28da7e88fd19f96ce9db340b)
[refactor: toPoint() 확장함수 사용으로 인한 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/af9f025f06fb059dfc1e678712f352dbb998dac3)

일단 toPoint()를 이용해서 라이브러리를 의존하고 있었기 때문에 Position을 Point로 변경해줄 수 있도록 Mapper라는 패키지를 만들어주었습니다. 

해당하는 패키지 내에서 확장함수로 toPoint()를 만들어 Board에서 import하여 사용할 수 있도록 구현해주었습니다!

lee-ji-hoon commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,90 @@
+                Violation.DOUBLE_THREE -> throw Exception(ERROR_DOUBLE_THREE)
+                Violation.DOUBLE_FOUR -> throw Exception(ERROR_DOUBLE_FOUR)
+                Violation.OVERLINE -> throw Exception(ERROR_OVERLINE)
+                Violation.NONE -> {}
```
만약 게임의 룰이 변경이 되어서 
```
3-3 반칙을 시도하려고 했을 때 해당 유저의 턴은 종료시킨다.
```
라는 것이 생겼다고 했을 때 이 코드를 View에서는 어떻게 처리해야 할까요?

```kotlin
private fun stoneAddedBoard(board: Board): Board {
    val result =
        runCatching {
           ...
        }.getOrElse { exception ->
            // 이렇게 처리할 수 밖에 없는 상황 같은데 상태에 대해서 다시 고민해봅시다.
            if (exception.message == "3-3 반칙이 발생했습니다") {
                ...
            }
        }

    return result
}
```

gahyunkim replied:
[feat: ViolationType Sealed class 생성](https://github.com/woowacourse/kotlin-omok/pull/115/commits/388dea3bffa16005875405ccf5627ced33c86353)
[refactor: ViolationType 구분 추가](https://github.com/woowacourse/kotlin-omok/pull/115/commits/96327ebcb6c245f86695d0f921b016435d4f39a9)
[refactor: ViolationType 적용 및 코드 구조화](https://github.com/woowacourse/kotlin-omok/pull/115/commits/53472572b26d0b50d9757a11fb5522140d634c72)

위와 같은 상황을 방지하기 위해서 Sealed Class를 만들어주어서 각각의 Exception을 관리할 수 있도록 해주었습니다. 또한, Violation 자체가 라이브러리에 의존하고 있기 때문에 model내에서의 의존성을 줄여주기 위해 mapper 클래스 내부에서 생성해주어 controller에서 사용하도록 구현해주었습니다.

lee-ji-hoon commented on src/main/kotlin/omok/model/stone/Stone.kt:
```diff
@@ -0,0 +1,10 @@
+    val position: Position,
+    val stoneState: StoneState,
+) {
+    constructor(lastStone: Pair<Position, StoneState>) : this(lastStone.first, lastStone.second)
```
이 부생성자는 단순 테스트만을 위해서 만들어진 부 생성자로 보이는데 맞을까요 🤔

gahyunkim replied:
[refactor: 필요없는 부생성자 삭제](https://github.com/woowacourse/kotlin-omok/pull/115/commits/48e82762150c938243f4ae8cd77032b9e510af1c)

단순 테스트를 위해서 만들었지만, 코드 리팩토링을 하면서 사용이 필요없어져서 삭제해주었습니다!

lee-ji-hoon commented on src/test/kotlin/omok/model/BoardTest.kt:
```diff
@@ -0,0 +1,28 @@
+import org.junit.jupiter.api.assertDoesNotThrow
+import org.junit.jupiter.api.assertThrows
+
+class BoardTest {
```
`placeStone` 에서 throw를 던지고 있는 부분들이 많은데 이 부분은 테스트가 필요 없다고 판단해서 테스트를 하지 않은 걸까요? 만약 그렇다면 어떤 이유로 테스트가 필요 없다고 판단했는지 궁금하네요

gahyunkim replied:
[test: Sealed class 및 BlackRenjuRule 확인을 위한 테스트 코드 구현](https://github.com/woowacourse/kotlin-omok/pull/115/commits/02db3e68734ab23a9e1b5454cc59f4caae40e15a)

BlackRenjuRule의 경우 라이브러리에서 테스트를 이미 진행하고 있기 때문에 placeStone()에서는 라이브러리를 믿고 테스트를 하지 않아도 되지 않을까라는 생각을 한 것 같습니다. 

하지만, 새롭게 Sealed클래스를 만들어줌에 따라 해당하는 throw가 제대로 발생하는 지 확인하기 위해 테스트코드를 작성해주었습니다!

lee-ji-hoon replied:
네 라이브러리가 이미 로직을 잘 실행해주고 있었지만, 이전 로직에서는 `어떠한 상태에 따라서 exception을 발생`하고 있기 때문에 필요하다고 보였습니다.

나중에 어떠한 라이브러리로 변경이 되어도 RenjuRule 자체는 동일하기에, 라이브러리만 변경이 되었을 뿐 오목 돌들로 Rule을 확인 했을 때 동일한 throw의 message가 있는지 판단을 하기 위해 필요하지 않았을까요?

> 물론 이미 잘 만들어주셨지만, 나중에 라이브러리를 쓴다고 테스트가 꼭 필요 없는 것은 아니며 그 값으로 어떠한 행동을 하느냐에 따라서 테스트가 필요해질 수 있다는 점을 기억해주시면 좋을 것 같네요.

gahyunkim replied:
너무 라이브러리에만 의존하지 말고, throw에 대한 테스트코드를 더 꼼꼼히 만들 수 있도록 하겠습니다!

lee-ji-hoon commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,90 @@
+                }
+        }
+
+    fun placeStone(nextPosition: Position): Board {
```
`placeStone` 을 보면 단순히 Store을 place만 할 것이라고 생각을 했는데 렌쥬룰도 확인 중인데 어 함수화를 좀 해보면 어떨까요.

> 객체지향 생활 규칙을 잊지 말고 항상 적용해보려고 노력해봅시다.

gahyunkim replied:
[refactor: Game 내부의 place() 및 세부 코드 수종](https://github.com/woowacourse/kotlin-omok/pull/115/commits/1c5a2be822fe655f77d4ff283d372f2812dcd0ae)

placeStone에서 Stone을 place하는 것뿐만 아니라 렌주룰도 확인하고 있다는 것을 알 수 있었습니다. 렌주룰의 로직을 직접적으로 알게하기 보다는 검사만 하도록 하고 내부 로직은 모르게 하기 위해서 구분해주었습니다! 

```kotlin
fun place(position: Position): Board {
        validateMove(position)
        applyMove(position)
        return board
}

private fun validateMove(position: Position) {
    currentRule().validate(board.stonesMap, position, turn)
}

private fun applyMove(position: Position) {
    board = board.positionAt(position, turn)
    lastStone = Stone(position, turn)
    turn = turn.next()
}

```

lee-ji-hoon commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,90 @@
+                }.map { (position, _) ->
+                    position.toPoint()
+                }
+        }
```
불변성을 신경 써주신 부분들 잘 확인했습니다.

그런데 문득 든 생각이 Black/White Points들이 변수로 선언이 되어 있어야 할 이유가 있을까?라는 생각이 드네요.

뭔가 억지로 변수로 빠지면서 불변성을 유지하게 만들어진 것 같은데 이 부분에 대해서는 어떻게 생각하는지 조이의 의견이 궁금하네요

gahyunkim replied:
[refactor: blackPoints, whitePoints를 변수에서 함수로 변경](https://github.com/woowacourse/kotlin-omok/pull/115/commits/da5b05e41fc1222f472871cc162e48f1ef51f7f0)

코니의 의견에 동의합니다! 저는 단순히 blackPoints와 whitePoints가 단순히 필터링된 리스트처럼 느껴져서 val로 선언했던것 같습니다. 그러다 보니 다른 개발자들이 보기에는 상태로 느껴질 것 같네요!

하지만 실제로는 상태가 아니라 계속 변하는 Map에서 계산한 결과이기 때문에 fun으로 만들어주는 것이 더 적절하다고 판단해 변경해주었습니다.

그리고 위에서 말씀해주신 라이브러리 의존성이 해당 코드에서도 사용되고 있어서 `List<rule.wrapper.point.Point>`를 `List<Position>`으로 변경해주고 확장함수를 이용해 Point로 변경해줄 수 있도록 해주었습니다.

lee-ji-hoon commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,90 @@
+        get() {
+            lastStone?.let { return omokRule.isLastStoneOmok(stonesMap, lastStone) }
+            return false
+        }
```
마지막 돌이 오목인지 아닌지 판단하는 변수가 있는데 아래와 같은 흐름으로도 생각해 볼 수 있을 것 같아요.

```
1. 돌을 놓는다.
2. 돌을 놓고 난 결과로 지금 오목인지 아닌지 상태를 반환한다.
```

그리고 이러한 여러 가지 상황들을 상태로 만들어볼 수도 있을 것 같네요

gahyunkim replied:
[feat: Rule 인터페이스 생성](https://github.com/woowacourse/kotlin-omok/pull/115/commits/5dd92dc88ea44510a5f720dd12e8242ee4dde869)
[feat: BlackRenjuRule 의존성을 줄이기 위한 BlackRuleChecker 구현](https://github.com/woowacourse/kotlin-omok/pull/115/commits/23a95dbad46c4f6b1d9c5b26b719051513f28b1e)

Rule이라는 인터페이스를 만들어준 후에 BlackOmokRule과 WhiteOmokRule을 각각의 구현체로 구현주었습니다. 

따라서 Rule에서는 마지막에 들어온 돌을 바탕으로 isWin()을 할 수 있게 되어서, 돌을 놓고 나서 변경된 stonesMap 즉, Board의 상태를 바탕으로 오목인지 아닌지를 Boolean으로 반환할 수 있도록 구현해주었습니다!

lee-ji-hoon commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,90 @@
+class Board private constructor(
+    val stonesMap: Map<Position, StoneState> = emptyMap(),
+    val lastStone: Stone? = null,
+) {
```
객체의 역할과 책임에 대해서 고민을 하기 전 `Board` 가 생성이 될 때 왜 마지막 돌이 필요할까요? 

`Board`가 갖고 있어야 할 상태인지 아니면 `Board` 가 생성이 될 때 필요한 것인지 다시 고민을 해보면 좋을 것 같네요.

> 그게 아니라면 게임을 진행하면서 계속 Board가 재생성이 되는게 맞을까? 도 고민해 보면 좋습니다.

gahyunkim replied:
[feat: Game 클래스 생성](https://github.com/woowacourse/kotlin-omok/pull/115/commits/e184d86420d0cbd649bb8b8a2a2503bd08924f29)

- Board: 단순히 오목판과 오목판 내의 Position의 상태만 관리함
- Game: 오목 게임의 진행과정 및 오목을 두는 과정을 관리함 

위와 같이 각 객체의 책임과 역할을 분리함으로써, Board내에서는 마지막 돌을 알 필요가 없다는 것을 알 수 있었습니다. 대신 게임을 진행하고 있는 Game 내에서 마지막 돌을 관리하도록 변경해주었습니다. 

게임을 진행하면서 Board를 계속해서 재 생성해준 이유는 Board 자체의 불변성을 유지하기 위해서였습니다. Board가 계속 생성된다는 단점은 있지만, 그와 별개로 불변성을 유지할 수 있기 때문에 불변성에 집중한다면 좋은 선택이 아닐까 하는 생각이 듭니다.

lee-ji-hoon replied:
```
불변성을 지킨다.
```

라는 관점에서는 잘해주셨습니다. `단 Board의 불변을 위해서 계속 재생성을 했을 때 어떤 이점이 있었을까?` 라는 고민은 좀 드네요. `Board`에 `place`라는 함수를 통해 Stone과 Position을 전달해 주고 아래처럼 할 수도 있었을 것 같은데요. 

```kotlin
class Board private constructor() {
    private val _stonesMap: MutableMap<Position, StoneColor> = mutableMapOf()
    val stonesMap: Map<Position, StoneColor> = _stonesMap.toMap()
```

이 방법이 Board라는 객체에 대해서 조금 더 유연하고, 자체적으로 할 일을 직접 할 수 있게 될 텐데요. 그러면 마지막으로 질문을 드리면 `Board` 자체를 계속 재할당하면서 불변을 유지하는 것이 어떠한 이점이 있었을까요?

gahyunkim replied:
[refactor: Board 내에서 가변 stonesMap 사용을 위한 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/0eb6f1a5795cfc28a8e77db39b81d8bf04f3f890)

위와 같은 방식을 사용하면 계속해서 보드를 생성하지 않고도 불변성을 유지할 수 있을 것 같습니다. 페어와 함께 고민했었던 부분이긴 하지만, 보드를 계속해서 재 생성하는 것 자체에서 큰 메모리상의 문제가 없을 것이라고 판단하기도 했고, 페어와 함께 프로그래밍 하는 과정에서 '엘강오' 책 이야기를 듣고 구현하다보니 보드를 재생성하는 코드가 나왔던 것 같습니다.

위와 같은 방식으로 변경해줌으로써 불필요하게 보드 객체를 계속해서 재생성하지 않고 _stonesMap을 클래스 내부에서만 조작할 수 있도록 해주었습니다.!

lee-ji-hoon commented on src/main/kotlin/omok/model/rule/OmokRule.kt:
```diff
@@ -0,0 +1,65 @@
+
+class OmokRule(
+    private val boardSize: Int,
+) {
```
`OmokRule` 은 오목에 대한 전반적을 `Rule` 의 역할을 갖고 있을 것이라고 보여서 해당 비즈니스로직은 전부 여기 있을 것이라고 예상을 했지만 실제로는 파편화가 되어 있는 듯한 느낌을 줍니다.

그 이유는 `Board` 에서 `Rule` 을 판단하고 있기 때문인데요.

```
어떻게 하면 `Rule` 에 대한 응집도를 높일 수 있을지 고민해 봅시다.
```

gahyunkim replied:
[feat: Rule 인터페이스 생성](https://github.com/woowacourse/kotlin-omok/pull/115/commits/5dd92dc88ea44510a5f720dd12e8242ee4dde869)

Rule에 대한 응집도를 높이기 위해서 Interface Rule()을 구현하여 WhiteOmokRule과 BlackOmokRule이 사용할 수 있는 규칙을 정의해주고, 각각의 Rule들에서 해당하는 로직을 구현할 수 있도록 해주었습니다!

lee-ji-hoon reviewed (changes requested):
안녕하세요 조이! 오목 리뷰어 코니입니다.

State라는 것을 학습하고 사용을 해보려고 한 흔적은 보이지만 아쉬운 부분들이 있으니 해당 부분에 대해서 한 번 고민을 해봤으면 좋겠네요.

그리고 객체의 역할과 책임에 대해서 고민에 대한 것과 전반적인 의존성 및 응집도에 대해서도 리뷰를 남겼습니다.

마지막 미션인 만큼 힘들겠지만 힘내봅시다!

gahyunkim committed:
refactor: parsing 위치 이동 및 boarSize 확장성 적용

gahyunkim committed:
refactor: StoneState를 StoneColor로 변경

gahyunkim committed:
refactor: StoneColor로 변경으로 인한 model 코드 수정

gahyunkim committed:
test: StoneColor로 변경으로 인한 test 코드 수정

gahyunkim committed:
refactor: StoneColor로 변경으로 인한 view 코드 수정

gahyunkim committed:
test: PositionTest 테스트 코드 삭제

gahyunkim committed:
feat: Point 라이브러리 의존을 줄이기 위한 확장함수 구현

gahyunkim committed:
refactor: toPoint() 확장함수 사용으로 인한 코드 수정

gahyunkim committed:
refactor: blackPoints, whitePoints를 변수에서 함수로 변경

gahyunkim committed:
feat: Game 클래스 생성

gahyunkim committed:
test: GameTest 테스트 코드 구현

gahyunkim committed:
refactor: Board의 책임과 역할 구분을 위한 코드 수정

gahyunkim committed:
feat: next()로 다음 턴을 변경하는 코드 구현

gahyunkim committed:
test: BoardTest 테스트 코드 변경

gahyunkim committed:
refactor: Board 변경 및 Game 클래스 생성으로 인한 코드 수정

gahyunkim committed:
chore: PositionToPoint 확장함수 삭제

gahyunkim committed:
chore: OmokRule 클래스 삭제

gahyunkim committed:
test: OmokRuleTest 삭제

gahyunkim committed:
feat: Rule 인터페이스 생성

gahyunkim committed:
feat: BlackRenjuRule 의존성을 줄이기 위한 BlackRuleChecker 구현

gahyunkim committed:
feat: 라이브러리 의존성을 줄이기 위한 PointMapper 구현

gahyunkim committed:
feat: Black,White OmokRule을 관리하는 클래스들 구현

gahyunkim committed:
test: WhiteOmokRuleTest 테스트 코드 구현

gahyunkim committed:
refactor: Game 내부의 place() 및 세부 코드 수종

gahyunkim committed:
test: GameTest 테스트 코드 수정

gahyunkim committed:
refactor: OmokControl 코드 수정

gahyunkim committed:
feat: ViolationType Sealed class 생성

gahyunkim committed:
refactor : 사용하지 않는 상수 삭제

gahyunkim committed:
refactor: ViolationType 구분 추가

gahyunkim committed:
refactor: ViolationType 적용 및 코드 구조화

gahyunkim committed:
refactor: 필요없는 부생성자 삭제

gahyunkim committed:
test: Sealed class 및 BlackRenjuRule 확인을 위한 테스트 코드 구현

lee-ji-hoon commented on src/main/kotlin/omok/mapper/BlackRuleChecker.kt:
```diff
@@ -0,0 +1,30 @@
+
+class BlackRuleChecker(
+    private val rule: BlackRenjuRule,
+    private val mapper: (Position) -> Point,
```
지금 이 mapper는 어떤 이유로 생성자로 받고 있는 것일까요? 굳이 필요해 보이지 않는 생성자로 보이는데 이렇게 했을 때 어떤 이점이 있을 것이라고 생각하셨을까요?

gahyunkim replied:
둘리의 라이브러리에 최대한 의존하지 않기 위해 BlackRuleChecker를 설계하게 되었습니다.
해당 라이브러리는 Point 타입을 사용하지만, 제 도메인에서는 Position을 사용하기 때문에, 이를 변환해주기 위한 mapper를 주입받는 구조로 구성했습니다.

이 방식을 사용하면 Point라는 외부 타입을 mapper라는 패키지의 BlackRuleChecker 내부에서만 사용하게 되어, Controller에서는 Point를 직접 알 필요가 없기 때문에 좋은 방법이라고 생각했습니다. 

실제로 Controller에서는 아래와 같이 간단히 mapper만 주입해주면 됩니다.

```kotlin
private val blackRuleChecker = BlackRuleChecker(
    rule = BlackRenjuRule(),
    mapper = { position -> PointMapper().from(position) },
)
```

이처럼 도메인 계층이 외부 라이브러리와의 결합 없이 동작하게 되므로,
라이브러리 변경에도 유연하고 테스트 또한 수월해지는 구조가 된다고 생각했습니다.

또한 Point로의 변환은 model이 아닌 mapper 패키지에 위치시켜
도메인 계층에서는 외부 구현 세부사항을 전혀 알지 않아도 되도록 설계했습니다.

lee-ji-hoon replied:
네 좋습니다. 그러면 아래와 같이 interface로 하면 여러 가지 Mapper를 상속받아서 구현을 하고 Map을 하는 것은 해당 객체에게 위임을 할 수도 있습니다.

```kotlin
fun interface Mapper {
    fun map(position: Position): Point
}
```

그러면 Mapper를 람다로 받았을 때는 어느 장점이 있다고 생각하시나요?

gahyunkim replied:
함수형 인터페이스는 함수를 하나만 갖는 인터페이스이고, 람다로 쉽게 사용할 수 있게 만들어줍니다. 

따라서 fun interface를 쓰면 람다로 좀 더 간결하게 작성이 가능하고, 필요할 때마다 다른 값을 넣어서 테스트를 편하게 한다거나, 다양한 코드를 사용할 수 있을 것 같습니다!

lee-ji-hoon commented on src/main/kotlin/omok/mapper/BlackRuleChecker.kt:
```diff
@@ -0,0 +1,30 @@
+            Violation.OVERLINE -> throw OverlineViolation
+            Violation.NONE -> return
+        }
+    }
```
`checkFoul` 에서 error면 무조건 throw를 던지고 그게 아니라면 return을 하는데 throw를 던져야만 했던 상황은 아닌 것 같은데요. throw를 사용해야겠다고 판단하는 조이의 기준이 궁금하네요.

> `return으로 어떠한 상태`를 넘기는건 안됐을까요?

gahyunkim replied:
[refactor: ViolationType 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/928da4d67e6bff08df141210fb764149e85c8456)
[refactor: ViolationType 수정으로 인한 Rule 인터페이스 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/353c494402c24fa0c3a35b898b9f925cd8126227)
[refactor: ViolationType 수정으로 인한 게임 진행 과정 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/8dc325774e6bfd5620121d61593e78e70ee9a554)

throw를 사용하게 되면 try-catch를 이용해서 어떤 반칙인지 판단하고 이를 가지고 게임을 진행해나갈 수 있기 때문에 throw를 사용했던 것 같습니다. 

throw 대신 return을 이용해서 각각의 DoubleThreeViolation, DoubleFourViolation과 같은 상태를 반환해주도록 하여 불필요하게 throw가 발생하지 않도록 구현해주었습니다.

violationType을 반환하도록 하여 해당하는 violation이 일어나는 경우에만 조건에 맞는 행동을 하도록 해주었습니다!

lee-ji-hoon commented on src/main/kotlin/omok/mapper/ViolationType.kt:
```diff
@@ -0,0 +1,11 @@
+    message: String,
+) : Exception(message)
+
+object DoubleThreeViolation : ViolationType("3-3 반칙이 발생했습니다")
```
throw 부분의 리뷰에 따라서 이것은 달라질 수도 있지만 

<img width="529" alt="image" src="https://github.com/user-attachments/assets/65cc38d2-1a56-4183-8eeb-e09b98e5f817" />

<img width="509" alt="image" src="https://github.com/user-attachments/assets/8f2c8a54-1a25-49d0-baee-6d6ea990cc53" />

이런 경고들이 왜 뜨는지 알아보는 것도 좋습니다.

gahyunkim replied:
[refactor: ViolationType 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/928da4d67e6bff08df141210fb764149e85c8456)

   
**readResolve**

kotlin의 object는 싱글턴이기 때문에 Java 직렬화 시스템과의 호환을 위해 readResolve()가 필요하게 됩니다. K2 컴파일러의 도입으로 이 내용이 더 강화되어서 위와 같은 표시가 나타날 수 있다고 합니다.

----

**data object**
object와 동일하지만, data class 처럼 내부적으로 toString(), equals(), hashCode()를 구현하고 있습니다.

- object는 로그를 남기게 되면 toString()에서 메모리 주소와 함께 전체 package의 패스가 남게 됩니다.
- 하지만, K2 compiler가 적용되면서 data object를 사용하면 toString()이 오버라이딩 되어있어 로그가 간단하게 나오는 것을 알 수 있었습니다. 


여기서 나타난 오류를 해결하기 위해서 object를 data object로 변경하여 구현해주었습니다!

lee-ji-hoon commented on src/main/kotlin/omok/model/game/Game.kt:
```diff
@@ -0,0 +1,56 @@
+        return board
+    }
+
+    private fun validateMove(position: Position) {
```
전반적으로 Move 보단 더 좋은 네이밍이 있을 것 같네요. 지금은 움직이는 것은 아니니까요!

gahyunkim replied:
[refactor: 불명확한 함수명 변경](https://github.com/woowacourse/kotlin-omok/pull/115/commits/d61693b4e9b9bc5859afd6d920e30bc7b16c87bb)

전체적으로 Game에 있는 함수들의 네이밍이 애매하게 느껴져서 각각 placeStone(), validatePosition, applyPlacement로 변경해주었습니다!

- placeStone(): 돌을 착수하는 함수 
    - 아래의 validatePosition과 applyPlacement를 모두 포함하고 있음
- validatePosition(): 현재 위치가 validate인지 파악하는 함수
    - validateMove-> validatePosition으로 변경해줌
- applyPlacement(): Move대신 Placement 즉, 배치를 적용해준다는 의미 적용
    - 변경되는 내용들에 대해서 새롭게 적용해주는 함수의 역할을 명확히 함

```kotlin
fun placeStone(position: Position): Board {
        validatePosition(position)
        applyPlacement(position)
        return board
}

private fun validatePosition(position: Position)      
    currentRule(turn).validate(board, position, turn)
}

private fun applyPlacement(position: Position) {
    board = board.positionAt(position, turn)
    lastStone = Stone(position, turn)
    turn = turn.next()
}

lee-ji-hoon commented on src/main/kotlin/omok/model/rule/BlackOmokRule.kt:
```diff
@@ -0,0 +1,37 @@
+        blackRuleChecker.checkFoul(positions.first, positions.second, nextPosition)
+    }
+
+    private fun positions(stones: Map<Position, StoneColor>): Pair<List<Position>, List<Position>> {
```
Rule의 역할만 봤을 때 직접 Position들을 필터링에서 찾는 것이 아니라 돌을 갖고 있는 객체에 한테서 전달을 받는 것이 좀 더 명확하지 않을까요?

gahyunkim replied:
[refactor: bordSize 및 black,whiteStones 필터링 코드 추가 및 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/45d31d95c5883f16a496aff85cceb94d5f59931f)
[refactor: boardSize 변경 및 black,whiteStones 필터링 추가로 인한 model 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/94b6281ae84796d8fa811a054d2a8f26924d5197)

결국 돌을 가지고 있는 객체는 Board이기 때문에 board로 부터 get을 이용해서 필터링된 값들을 받을 수 있도록 구현해주었습니다!

lee-ji-hoon commented on src/main/kotlin/omok/model/rule/BlackOmokRule.kt:
```diff
@@ -0,0 +1,37 @@
+        if (color != StoneColor.BLACK) return
+
+        val positions = positions(stones)
+        blackRuleChecker.checkFoul(positions.first, positions.second, nextPosition)
```
Pair는 사용을 지양하는 것이 좋습니다.

지금은 구현했을 때 first/second가 무엇인지 명확하지만 당장 한 달만 지나고 해당 값이 무엇인지 헷갈리기도 하고, 타인이 봤을 때 동일한 `List<Position>` 이기에 이해하기도 어렵습니다.

gahyunkim replied:
[refactor: bordSize 및 black,whiteStones 필터링 코드 추가 및 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/45d31d95c5883f16a496aff85cceb94d5f59931f)
[refactor: boardSize 변경 및 black,whiteStones 필터링 추가로 인한 model 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/94b6281ae84796d8fa811a054d2a8f26924d5197)

Pair를 사용하면서 실제로 first와 second가 무엇인지 몰라서 헷갈렸던 적이 있었어서 코니의 의견에 매우 동의합니다..!

Pair대신 위에서 말씀해주신 리뷰처럼 stones를 가지고 있는 객체로부터 get을 이용해서 필터링된 값들을 받을 수 있도록 구현해주었습니다.

lee-ji-hoon commented on src/main/kotlin/omok/model/rule/Rule.kt:
```diff
@@ -0,0 +1,18 @@
+import omok.model.stone.position.Position
+
+interface Rule {
+    fun isWin(
```
Rule Interface 좋습니다. 단 Rule이라는 너무 추상적인 단어보다는 정확히 어떠한 Rule에 대한 Interface인지 접두사를 잘 붙여주는 것도 좋을 것 같네요.

gahyunkim replied:
[refactor: OmokRule로 인터페이스 명 변경](https://github.com/woowacourse/kotlin-omok/pull/115/commits/a80d895c75ade1d8e85d66594e409700d61ab1cc)

Rule이라는 인터페이스 이름을 좀 더 명확하게 하기 위해 OmokRule로 변경해주었습니다.

lee-ji-hoon reviewed (changes requested):
조이 안녕하세요! 리뷰 반영 및 의견 남겨주신 부분 잘 확인했습니다.

model이 외부 의존이 없게끔 구현 잘해주셨습니다.  

객체에 대한 책임과 각 레이어 별로 어떠한 일을 해야 하는지는 블랙잭까지 많이 해보았기에 조이가 깊게 고민해 볼 만한 주제를 몇 개 드렸습니다. 해당 리뷰들에 대해서 고민을 해보고 조이의 의견 및 반영을 해주시고 1단계 마지막 리뷰 요청을 주시면 되겠습니다.

다음 리뷰 요청 이후에는 바로 2단계를 진행할 예정이니 요청 이후에는 3,4단계를 진행하고 계셔도 됩니다.

gahyunkim committed:
test: PositionFixture 구현

gahyunkim committed:
test: fixture 적용

gahyunkim committed:
feat: BoardDimensions data class 생성

gahyunkim committed:
refactor: Position에서의 검증 삭제

gahyunkim committed:
refactor: blackRenjuRule에서의 Point 매핑 오류 해결

gahyunkim committed:
refactor: bordSize 및 black,whiteStones 필터링 코드 추가 및 수정

gahyunkim committed:
refactor: boardSize 변경으로 인한 코드 수정 및 검증 추가

gahyunkim committed:
refactor: boardSize 변경으로 인한 코드 수정

gahyunkim committed:
refactor: boardSize 변경 및 black,whiteStones 필터링 추가로 인한 model 코드 수정

gahyunkim committed:
test: boardSize 변경으로 인한 코드 수정 및 테스트 코드 추가

gahyunkim committed:
test: PositionTest 테스트 코드 삭제

gahyunkim committed:
test: boardSize 및 black,whiteStones 핉링 코드 추가로 인한 테스트 코드 수정

gahyunkim committed:
refactor: OmokRule로 인터페이스 명 변경

gahyunkim committed:
refactor: 불명확한 함수명 변경

- place() -> placeStone()
- validateMove() -> validatePosition()
- applyMove() -> applyPlacement()

gahyunkim committed:
test: 불명확한 함수명 변경으로 인한 테스트 코드 수정

gahyunkim committed:
refactor: ViolationType 수정

gahyunkim committed:
refactor: ViolationType 수정으로 인한 Rule 인터페이스 수정

gahyunkim committed:
refactor: ViolationType 수정으로 인한 게임 진행 과정 코드 수정

gahyunkim committed:
refactor: ViolationType 수정으로 인한 테스트 코드 수정

gahyunkim committed:
refactor: Board 내에서 가변 stonesMap 사용을 위한 코드 수정

lee-ji-hoon commented on src/main/kotlin/omok/model/board/Board.kt:
```diff
@@ -0,0 +1,40 @@
+class Board private constructor(
+    private val dimensions: BoardDimensions,
+) {
+    private val _stonesMap: MutableMap<Position, StoneColor> = mutableMapOf()
+    val stonesMap: Map<Position, StoneColor>
+        get() = _stonesMap.toMap()
```
```kotlin
private val _stonesMap: MutableMap<Position, StoneColor> = mutableMapOf()
val stonesMap: Map<Position, StoneColor> = _stonesMap.toMap()
```

지금의 구조에서 `get()` 을 했을 때와 단순히 `=` 할당을 했을 때 어떤 차이점이 있을까요? 어떤 이점을 보고 `get()` 을 선택했는지 궁금하네요!

gahyunkim replied:
=으로 할당하게 되면 _stonesMap이 생성될 당시의 상태가 stonesMap에 그대로 고정되지만,
get()을 사용하면 _stonesMap이 변경되더라도 항상 최신 상태를 외부에서 참조할 수 있습니다!

그래서, 외부에서 최신 상태의 보드를 사용할 수 있도록 stonesMap은 get()을 활용해 정의했습니다.

lee-ji-hoon commented on src/main/kotlin/omok/controller/OmokControl.kt:
```diff
@@ -0,0 +1,61 @@
+        } else {
+            turn()
+        }
+    }
```
이 trun이라는 함수가 좀 가혹할 정도로 많은 일을 하고 있는 것 같네요 

turn 뿐만 아니라 이 게임의 흐름 전체를 하나의 함수에서 처리하고 있는데 함수화를 좀 해보면 어떨까요?

gahyunkim replied:
turn에서 너무 많은 일을 하고 있는 것 같아서 함수로 나눠서 turn의 역할을 조금 줄여주었습니다. 

turn이 아니라 게임이 play된다고 생각해서 일단 함수명을 play로 변경해주고, 아래처럼 구현해주어 에러 상태에 따른 행동을 하고 게임을 자연스럽게 진행할 수 있도록 구현해주었습니다.

```kotlin
private fun play(showBoard: Boolean = true) {
    if (showBoard) printCurrentState()

    val position = readPosition() ?: return play(showBoard = false)
    val violation = processPlacement(position)

    if (violation != PlacementError.NoViolation) {
        handleViolation(violation)
        return
    }

    if (checkGameEnd()) return

    play()
}
```

lee-ji-hoon commented on src/main/kotlin/omok/mapper/ViolationType.kt:
```diff
@@ -0,0 +1,13 @@
+    val message: String,
+)
+
+data object DoubleThreeViolation : ViolationType("3-3 반칙이 발생했습니다")
+
+data object DoubleFourViolation : ViolationType("4-4 반칙이 발생했습니다")
+
+data object OverlineViolation : ViolationType("장목 반칙이 발생했습니다")
+
+data object NoViolation : ViolationType("반칙이 발생하지 않았습니다")
```
네 이제 Android로 넘어가기 전 이젠 콘솔로 하는 것이 아닌 실제 View를 활용하게 될 테니 고민해 볼 만한 리뷰 하나 남겨드리겠습니다.

지금은 Console에 나올 메시지를 Board가 정해서 넘겨주고 있는데요. 

1. View에 대한 Message를 Model에서 정해주고 있는 것은 아닐까?
2. 지금 왜 상태로 나눴을까?

이 2가지를 고민해 보시면 좋을 것 같네요 🤔

gahyunkim replied:
> View에 대한 Message를 Model에서 정해주고 있는 것은 아닐까?

라는 코니의 의문을 보니 에러 메시지를 도메인에서 만들어주고 있다는 생각이 들었습니다. 그래서 메시지를 담아주지 않고 return 값으로 상태만 반환한 뒤, 해당하는 상태에 따라 메시지를 outputView에서 지정해줄 수 있도록 새롭게 구현해주었습니다!

> 지금 왜 상태로 나눴을까?

상태로 나누게 되니, 사용자가 선택을 잘못하거나 3-3반칙과 같은 게임 진행에 오류가 생겼을 때 상태에 따라 원하는 행동을 하도록 만들어줄 수 있어서 좋은 것 같습니다!
또한, 도메인에서의 분기를 만들어주어서 게임 진행이 더 자연스러워질 수 있을 것 같습니다.

lee-ji-hoon commented on src/main/kotlin/omok/model/game/Game.kt:
```diff
@@ -0,0 +1,59 @@
+
+    fun getTurn(): StoneColor = turn
+
+    fun getLastStone(): Stone? = lastStone
```
이렇게 get을 따로 구현을 해야 했을까?라는 의문이 있네요.

var로 하되 private set을 해두는 것과 지금 방법은 어떠한 차이점이 있을까요?

gahyunkim replied:
[refactor: Game get 코드 수정](https://github.com/woowacourse/kotlin-omok/pull/115/commits/31e0206daf624bd22c4d82725ccf5b636848f541)
get이 아니라 private set을 이용하면, 내부에서는 자유롭게 사용하도록 하고 외부에서는 사용하지 못하도록 방지할 수 있을 것 같습니다!

코니의 제안이 더 코틀린스러운 방법인 것 같네요!

lee-ji-hoon commented on src/main/kotlin/omok/model/rule/WhiteOmokRule.kt:
```diff
@@ -0,0 +1,60 @@
+    override fun isWin(
+        board: Board,
+        lastStone: Stone,
+    ): Boolean = calculateTotalCount(board.stonesMap, lastStone) >= 5
```
오목에서 5라는 매직넘버는 꽤 중요한 값일 것 같은데요. 상수화를 해두면 좋을 것 같네요

gahyunkim replied:
[refactor: OMOK_DEFAULT_VALUE 값 상수화](https://github.com/woowacourse/kotlin-omok/pull/115/commits/b26c604efb03c33a5a7d54d79aef14aac3412e6c)

오목이 될 수 있는 기본 값을 OMOK_DEFAULT_VALUE로 상수화해주었습니다!

lee-ji-hoon reviewed (approved):
안녕하세요 조이! 빠르게 잘 리뷰를 반영해 주셨습니다.

조이가 3단계를 진행하기 전 마지막으로 확인해 볼 만한 사항들을 리뷰로 남겼으며, 지금 구현한 모델들이 Android에 얼마나 잘 녹아들지 아니면 무슨 문제가 생길지 직접 몸소 체험을 해볼 시간이 왔네요!

다음 단계는 Android로 구현을 하는 것이다 보니 이번 단계는 현재 PR에서 마무리하기로 했으니 머지를 하도록 하겠습니다.

1,2단계 고생했습니다

gahyunkim committed:
refactor: OMOK_DEFAULT_VALUE 값 상수화

gahyunkim committed:
refactor: Board에서 사용하지 않는 코드 삭제

gahyunkim committed:
refactor: Game get 코드 수정

gahyunkim committed:
refactor: Game get 코드 수정으로 인한 Control 수정

gahyunkim committed:
test: Game get 코드 수정으로 인한 테스트 코드 수정
